compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1097,12 +1102,69 @@ private boolean validateTableType(BTableType tableType, List<BLangRecordLiteral>
                             ((BIntersectionType) constraint).effectiveType,
                     tableType.keyPos);
 
-            return (isKeySpecifierValidated);
+            return (isKeySpecifierValidated && validateTableKeyValue(fieldNameList, recordLiterals));
+        }
+
+        return true;
+    }
+
+    private boolean validateTableKeyValue(List<String> keySpecifierFieldNames,
+                                                           List<BLangRecordLiteral> recordLiterals) {
+
+        for (String fieldName : keySpecifierFieldNames) {
+            for (BLangRecordLiteral recordLiteral : recordLiterals) {
+                BLangRecordKeyValueField recordKeyValueField = getRecordKeyValueField(recordLiteral, fieldName);
+                if (recordKeyValueField != null && isConstExpression(recordKeyValueField.getValue())) {
+                    continue;
+                }
+
+                dlog.error(recordLiteral.pos,
+                        DiagnosticErrorCode.KEY_SPECIFIER_FIELD_VALUE_MUST_BE_CONSTANT_EXPR, fieldName);
+                resultType = symTable.semanticError;
+                return false;
+            }
         }
 
         return true;
     }
 
+    private boolean isConstExpression(BLangExpression expression) {
+        switch(expression.getKind()) {
+            case LITERAL:
+            case NUMERIC_LITERAL:
+            case STRING_TEMPLATE_LITERAL:
+            case XML_ELEMENT_LITERAL:
+            case XML_TEXT_LITERAL:
+            case LIST_CONSTRUCTOR_EXPR:
+            case TABLE_CONSTRUCTOR_EXPR:
+            case RECORD_LITERAL_EXPR:
+            case TYPE_CONVERSION_EXPR:
+            case UNARY_EXPR:
+            case BINARY_EXPR:
+            case TYPE_TEST_EXPR:
+            case TERNARY_EXPR:
+                return true;
+            case SIMPLE_VARIABLE_REF:
+                return (((BLangSimpleVarRef) expression).symbol.tag & SymTag.CONSTANT) == SymTag.CONSTANT;
+            case GROUP_EXPR:
+                return isConstExpression(((BLangGroupExpr) expression).expression);
+            default:
+                return false;
+        }
+    }
+
+    private BLangRecordKeyValueField getRecordKeyValueField(BLangRecordLiteral recordLiteral,
+                                                            String fieldName) {
+        for (RecordLiteralNode.RecordField recordField : recordLiteral.fields) {
```

Comment:
cant we use streams here? we need to see if there is a performance issue using java streams